README.md
[![Build Status](https://travis-ci.org/mudler/Mojolicious-Plugin-ViewBuilder.png?branch=master)](https://travis-ci.org/mudler/Mojolicious-Plugin-ViewBuilder)
# NAME

Mojolicious::Plugin::ViewBuilder - a Mojolicious plugin that allows to chain templates generated by other plugins

# SYNOPSIS

    #Load the plugin

    # - Mojolicious
    $self->plugin('ViewBuilder');

    # - Mojolicious::Lite
    plugin 'ViewBuilder';

    # then, use it in your template, declare a new area, labeled "profile"
    <%= pluggable_view 'profile' %>

    # now, in the plugin, that will attach the new helper to the view

    sub register {
        my ( $self, $app, $conf ) = @_;

        $app->helper(
            test => sub {
                shift->render_to_string( "test", some_data => 1 );
            }
        );

        #attach the test helper to the view "profile"
        $app->add_view( "profile", "test" );

    }
    1;

    # or you can append a callback

    sub register {
        my ( $self, $app, $conf ) = @_;
        $app->add_view(
            profile => sub {
                shift->render_to_string( "test", some_data => 42 );
            }
        );
    }
    1;
    __DATA__
    @@ test.html.ep
    huuuray!
    % if( stash("some_data") and stash("some_data")==42){
     double it!
    % }

# DESCRIPTION

[Mojolicious::Plugin::ViewBuilder](https://metacpan.org/pod/Mojolicious::Plugin::ViewBuilder) is a [Mojolicious](https://metacpan.org/pod/Mojolicious) plugin. Chains templates generated by other plugins, associating them within the view. This is useful only if your workflow is to reap into little pieces the webapp (e.g. in plugins)

# METHODS

[Mojolicious::Plugin::ViewBuilder](https://metacpan.org/pod/Mojolicious::Plugin::ViewBuilder) inherits all methods from
[Mojolicious::Plugin](https://metacpan.org/pod/Mojolicious::Plugin) and implements the following new ones.

## pluggable\_view

    <%=pluggable_view "profile" %>

Will render all the attached plugin associated within the view

## add\_view

    $app->add_view("view","helper");
    $app->add_view(view => sub{shift->render_to_string("Hello mojolicious!")});

Attach the "helper" within the "view"

# LICENSE

Copyright (C) mudler.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# AUTHOR

mudler <mudler@dark-lab.net>


Changes
Revision history for Perl extension Mojolicious-Plugin-ViewBuilder

{{$NEXT}}

0.05 2015-01-26T22:53:55Z

    - Minor fixes

0.04 2014-11-19T22:23:10Z

    - POD fix

0.03 2014-11-19T22:15:14Z

    - POD fixes
    - now pluggable_view() pass the args at the attached subs

0.02 2014-11-19T21:54:55Z

    - now add_view() support code refs as argument

0.01 2014-09-07T12:17:47Z

    - first release



t/lib/Test/TestPlugin.pm
package Test::TestPlugin;

use Mojo::Base 'Mojolicious::Plugin';

sub register {
    my ( $self, $app, $conf ) = @_;
    push @{ $app->renderer->classes }, __PACKAGE__;
    $app->add_view( "activity", "test" );
    $app->add_view( "activity", "test2" );

    $app->helper(
        test => sub {
            return shift->render_to_string(
                template => "test",
                format   => "html"
            );
        }
    );
    $app->helper(
        test2 => sub {
            return shift->render_to_string(
                template => "test2",
                format   => "html"
            );
        }
    );

}

1;
__DATA__
@@ test.html.ep
double it!
@@ test2.html.ep
42

lib/Mojolicious/Plugin/ViewBuilder.pm
package Mojolicious::Plugin::ViewBuilder;
our $VERSION = "0.05";

use Mojo::Base 'Mojolicious::Plugin';

sub register {
    my ( $self, $app, $conf ) = @_;
    push @{ $app->renderer->classes }, __PACKAGE__;

    $app->helper(
        add_view => sub {
            shift;
            push( @{ $app->{views}->{ +shift() } }, shift );
        }
    );

    $app->helper(
        pluggable_view => sub {
            my $c=shift;
            die "No view defined" unless defined $_[0];
            my $output;
            $output .= ref $_ eq 'CODE' ? $_->( $c, $app,@_ ) : $app->$_($c,@_)
                for ( @{ $app->{views}->{ +shift() } } );
            return $output;
        }
    );

}

1;

__END__

=encoding utf-8

=head1 NAME

Mojolicious::Plugin::ViewBuilder - a Mojolicious plugin that allows to chain templates generated by other plugins

=head1 SYNOPSIS


  #Load the plugin

  # - Mojolicious
  $self->plugin('ViewBuilder');

  # - Mojolicious::Lite
  plugin 'ViewBuilder';

  # then, use it in your template, declare a new area, labeled "profile"
  <%= pluggable_view 'profile' %>

  # now, in the plugin, that will attach the new helper to the view

  sub register {
      my ( $self, $app, $conf ) = @_;

      $app->helper(
          test => sub {
              shift->render_to_string( "test", some_data => 1 );
          }
      );

      #attach the test helper to the view "profile"
      $app->add_view( "profile", "test" );

  }
  1;

  # or you can append a callback

  sub register {
      my ( $self, $app, $conf ) = @_;
      $app->add_view(
          profile => sub {
              shift->render_to_string( "test", some_data => 42 );
          }
      );
  }
  1;
  __DATA__
  @@ test.html.ep
  huuuray!
  % if( stash("some_data") and stash("some_data")==42){
   double it!
  % }

=head1 DESCRIPTION

L<Mojolicious::Plugin::ViewBuilder> is a L<Mojolicious> plugin. Chains templates generated by other plugins, associating them within the view. This is useful only if your workflow is to reap into little pieces the webapp (e.g. in plugins)

=head1 METHODS

L<Mojolicious::Plugin::ViewBuilder> inherits all methods from
L<Mojolicious::Plugin> and implements the following new ones.

=head2 pluggable_view

  <%=pluggable_view "profile" %>

Will render all the attached plugin associated within the view

=head2 add_view

  $app->add_view("view","helper");
  $app->add_view(view => sub{shift->render_to_string("Hello mojolicious!")});

Attach the "helper" within the "view"

=head1 LICENSE

Copyright (C) mudler.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

mudler E<lt>mudler@dark-lab.netE<gt>

=cut

